pkg/telemetry/cass_prometheus_resourcer.go

```diff
@@ -0,0 +1,459 @@
+    - regex: jvm\.gc\.(\w+)\.(\w+)
+      replacement: ${1}
+      sourceLabels:
+      - mcac
+      targetLabel: collector_type
+    - regex: jvm\.gc\.(\w+)\.(\w+)
+      replacement: mcac_jvm_gc_${2}
+      sourceLabels:
+      - mcac
+      targetLabel: __name__
+    - regex: jvm\.memory\.(\w+)\.(\w+)
+      replacement: ${1}
+      sourceLabels:
+      - mcac
+      targetLabel: memory_type
+    - regex: jvm\.memory\.(\w+)\.(\w+)
+      replacement: mcac_jvm_memory_${2}
+      sourceLabels:
+      - mcac
+      targetLabel: __name__
+    - regex: jvm\.memory\.pools\.(\w+)\.(\w+)
+      replacement: ${2}
+      sourceLabels:
+      - mcac
+      targetLabel: pool_name
+    - regex: jvm\.memory\.pools\.(\w+)\.(\w+)
+      replacement: mcac_jvm_memory_pool_${2}
+      sourceLabels:
+      - mcac
+      targetLabel: __name__
+    - regex: jvm\.fd\.usage
+      replacement: mcac_jvm_fd_usage
+      sourceLabels:
+      - mcac
+      targetLabel: __name__
+    - regex: jvm\.buffers\.(\w+)\.(\w+)
+      replacement: ${1}
+      sourceLabels:
+      - mcac
+      targetLabel: buffer_type
+    - regex: jvm\.buffers\.(\w+)\.(\w+)
+      replacement: mcac_jvm_buffer_${2}
+      sourceLabels:
+      - mcac
+      targetLabel: __name__
+    - regex: (mcac_.*);.*(_micros_bucket|_bucket|_micros_count_total|_count_total|_total|_micros_sum|_sum|_stddev).*
+      replacement: ${1}${2}
+      separator: ;
+      sourceLabels:
+      - __name__
+      - prom_name
+      targetLabel: __name__
+    - action: labeldrop
+      regex: prom_name`
+
+// mustLabels() returns the set of labels essential to managing the Prometheus resources. These should not be overwritten by the user.
+func (cfg CassPrometheusResourcer) mustLabels() map[string]string {
+	return map[string]string{
+		"app.kubernetes.io/managed-by": "k8ssandra-operator",
+		"app.kubernetes.io/part-of":    "k8ssandra",
+		"k8ssandra.io/cluster":         cfg.ClusterName,
+		"k8ssandra.io/datacenter":      cfg.DataCenterName,
+		"k8ssandra.io/monitors":        "cassandra-datacenter",
+	}
+}
+
+// NewServiceMonitor returns a Prometheus operator ServiceMonitor resource.
+func (cfg CassPrometheusResourcer) NewServiceMonitor() (promapi.ServiceMonitor, error) {
+	// validate the object we're being passed.
+	if cfg.CassandraNamespace == "" || cfg.ServiceMonitorName == "" || cfg.DataCenterName == "" || cfg.ClusterName == "" {
+		return promapi.ServiceMonitor{}, TelemetryConfigIncomplete{}
+	}
+	// Overwrite any CommonLabels the user has asked for if they conflict with the labels essential for the functioning of the operator.
+	var mergedLabels = map[string]string{}
+	for k, v := range cfg.CommonLabels {
+		mergedLabels[k] = v
+	}
+	mustLabels := cfg.mustLabels()
+	for k, v := range mustLabels {
+		mergedLabels[k] = v
+	}
+	var endpointHolder promapi.ServiceMonitor
+	decode := scheme.Codecs.UniversalDeserializer().Decode
+	_, _, err := decode([]byte(endpointString), nil, &endpointHolder)
```

Comment:
Since `endpointString` is static, couldn't we just didn't this at package initialization time in an `init` function? Actually, is it necessary at all at runtime? Couldn't this be done in a unit test?

Reply:
Change made, I don't particularly like what this does to error handling (we basically need to panic if something goes wrong with the static initialisation...) but if that's preferred, it will surely work.